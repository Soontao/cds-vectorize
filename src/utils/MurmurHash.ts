/* eslint-disable prefer-const */
/* eslint-disable camelcase */

/**
 * murmurhash3 private implementation
 *
 * @param plain
 * @param seed random seed
 * @returns the hashed value [0, 2^32 - 1]
 */
export function murmurhash3(plain: string, seed: number = 42): number {
  let remainder, bytes, h1, c1, c2, k1, i;

  remainder = plain.length & 3; // key.length % 4
  bytes = plain.length - remainder;
  h1 = seed;
  c1 = 0xcc9e2d51;
  c2 = 0x1b873593;
  i = 0;

  while (i < bytes) {
    k1 =
      (plain.charCodeAt(i) & 0xff) |
      ((plain.charCodeAt(++i) & 0xff) << 8) |
      ((plain.charCodeAt(++i) & 0xff) << 16) |
      ((plain.charCodeAt(++i) & 0xff) << 24);
    ++i;

    k1 = (k1 * c1) >>> 0;
    k1 = (k1 << 15) | (k1 >>> 17);
    k1 = (k1 * c2) >>> 0;

    h1 ^= k1;
    h1 = (h1 << 13) | (h1 >>> 19);
    h1 = (h1 * 5 + 0xe6546b64) >>> 0;
  }

  k1 = 0;

  switch (remainder) {
    case 3:
      k1 ^= (plain.charCodeAt(i + 2) & 0xff) << 16;
    case 2:
      k1 ^= (plain.charCodeAt(i + 1) & 0xff) << 8;
    case 1:
      k1 ^= plain.charCodeAt(i) & 0xff;

      k1 = (k1 * c1) >>> 0;
      k1 = (k1 << 15) | (k1 >>> 17);
      k1 = (k1 * c2) >>> 0;
      h1 ^= k1;
  }

  h1 ^= plain.length;

  h1 ^= h1 >>> 16;
  h1 = (h1 * 0x85ebca6b) >>> 0;
  h1 ^= h1 >>> 13;
  h1 = (h1 * 0xc2b2ae35) >>> 0;
  h1 ^= h1 >>> 16;

  return h1 >>> 0;
}

export default murmurhash3;
