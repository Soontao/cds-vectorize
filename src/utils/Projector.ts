/* eslint-disable @typescript-eslint/indent */
import { ObjectEncoder, ObjectMetadata } from "../index.js";

export function flattenObject(obj: any, parentKey?: string): Array<{ key: string; val: string }> {
  const result: Array<{ key: string; val: string }> = [];

  for (const [key, val] of Object.entries(obj)) {
    const currentKey = parentKey ? `${parentKey}_${key}` : key;

    if (typeof val === "object" && val !== null) {
      // Recursively flatten nested objects
      result.push(...flattenObject(val, currentKey));
    } else {
      if (["number", "string"].includes(typeof val)) {
        result.push({ key: currentKey, val: String(val) });
      } else {
        result.push({ key: currentKey, val: JSON.stringify(val) });
      }
    }
  }

  return result;
}

export function createProjectorValues(
  meta: ObjectMetadata,
  items: Array<any>,
  importantFeaturesIndexes?: Array<number>,
) {
  const e = new ObjectEncoder(meta);

  const vectors = items.map((i) => e.encode(i));
  const tsv = vectors
    .map((v) =>
      importantFeaturesIndexes == undefined
        ? v.map(String).join("\t")
        : importantFeaturesIndexes
            .map((idx) => v[idx])
            .map(String)
            .join("\t"),
    )
    .join("\n");
  const metaHeaders = flattenObject(items[0]).map((v) => v.key);
  const metadata = items.map((i) => flattenObject(i));
  const metaTsv = [
    metaHeaders.join("\t"),
    ...metadata.map((meta) =>
      Object.values(meta)
        .map((v) => v.val)
        .join("\t"),
    ),
  ].join("\n");
  return {
    tsv,
    metaTsv,
  };
}
