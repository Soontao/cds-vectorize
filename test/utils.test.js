import { randomUUID } from "crypto";
import { murmurhash3 } from "../src/utils/MurmurHash.js";
import { createProjectorValues, flattenObject } from "../src/utils/Projector.js";

describe("Utils Test Suite", () => {
  it("should return the correct hash value for a given key and seed", () => {
    expect(murmurhash3("Hello, MurmurHash!", 42)).toEqual(363275047);
    expect(murmurhash3("Jest is awesome!", 123)).toEqual(2017758788);
  });

  it("should return the same hash value for the same key and seed", () => {
    const key = "Test Key";
    const seed = 987;

    expect(murmurhash3(key, seed)).toBe(murmurhash3(key, seed));
  });

  it("should pass fuzzy test for murmurhash3", () => {
    for (const _ of Array(500).fill(0)) {
      const h = murmurhash3(randomUUID());
      expect(h).toBeGreaterThan(0);
      expect(h).toBeLessThan(2 ** 32 - 1);
    }
  });

  it('flattens a simple object', () => {
    const input = { a: 1, b: { c: 2 } };
    const expected = [
      { key: 'a', val: '1' },
      { key: 'b_c', val: '2' },
    ];

    const result = flattenObject(input);

    expect(result).toEqual(expected);
  });

  it('flattens an object with arrays', () => {
    const input = { a: { b: [{ c: 1 }, { c: 2 }] } };
    const expected = [
      { key: 'a_b_0_c', val: '1' },
      { key: 'a_b_1_c', val: '2' },
    ];

    const result = flattenObject(input);

    expect(result).toEqual(expected);
  });

  it('flattens an object with null values', () => {
    const input = { a: null, b: { c: 2 } };
    const expected = [
      { key: 'a', val: 'null' },
      { key: 'b_c', val: '2' },
    ];

    const result = flattenObject(input);

    expect(result).toEqual(expected);
  });

  it('creates projector values', () => {
    const meta = {
      properties: [
        {
          name: 'property1',
          type: 'category',
          values: ['value1', 'value2', 'value3']
        },
      ],
    };

    const items = [
      { property1: 'value1' },
      { property1: 'value2' },
    ];

    const result = createProjectorValues(meta, items);

    expect(result.tsv).toMatchSnapshot()
    expect(result.metaTsv).toMatchSnapshot();
  });
});
