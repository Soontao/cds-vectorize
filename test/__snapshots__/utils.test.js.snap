// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Utils Test Suite creates projector values 1`] = `
"1	0	0
0	1	0"
`;

exports[`Utils Test Suite creates projector values 2`] = `
"property1
value1
value2"
`;
